File: src/Scanner.cs
using Collections = System.Collections.Generic;
using IO = System.IO;
using Text = System.Text;

public sealed class Scanner
{
    private readonly Collections.IList<object> result;
    public Scanner(IO.TextReader input)
    {
        this.result = new Collections.List<object>();
        this.Scan(input);
    }

    public Collections.IList<object> Tokens
    {
        get { return this.result; }
    }

    #region ArithmithicConstants

    public static readonly object Add = BinOp.Add;
    public static readonly object Sub = BinOp.Sub;
    public static readonly object Mul = BinOp.Mul;
    public static readonly object Div = BinOp.Div;

    public static readonly object Equal = BinOp.Equal;
    public static readonly object Less = BinOp.Less;
    public static readonly object More = BinOp.More;
    public static readonly object LessOrEqual = BinOp.LessOrEqual;
    public static readonly object MoreOrEqual = BinOp.MoreOrEqual;

    public static readonly object Semi = new object();
    public static readonly object Assign = new object();

    public static readonly object PerenthStart = new object();
    public static readonly object PerenthEnd = new object();

    public static readonly object BraceStart = new object();
    public static readonly object BraceEnd = new object();

    public static readonly object SquareStart = new object();
    public static readonly object SquareEnd = new object();

    #endregion

    private void Scan(IO.TextReader input)
    {
        while (input.Peek() != -1)
        {
            char ch = (char)input.Peek();

            if (char.IsWhiteSpace(ch))
            {
                input.Read();
            }
            else if (char.IsLetter(ch) || ch == '_')
            {
                Text.StringBuilder accum = new Text.StringBuilder();
                while (char.IsLetter(ch) || ch == '_')
                {
                    accum.Append(ch);
                    input.Read();

                    if (input.Peek() == -1)
                    {
                        break;
                    }
                    else
                    {
                        ch = (char)input.Peek();
                    }


                }
                this.result.Add(accum.ToString());
            }
            else if (ch == '"')
            {
                Text.StringBuilder accum = new Text.StringBuilder();

                input.Read();

                if (input.Peek() == -1)
                {
                    throw new System.Exception("Unterminated string literal");
                }
                while ((ch = (char)input.Peek()) != '"')
                {
                    accum.Append(ch);
                    input.Read();

                    if (input.Peek() == -1)
                    {
                        throw new System.Exception("Unterminated string literal");
                    }

                }
                input.Read();
                this.result.Add(accum);
            }
            else if (char.IsDigit(ch))
            {
                Text.StringBuilder accum = new Text.StringBuilder();

                while (char.IsDigit(ch))
                {
                    accum.Append(ch);
                    input.Read();

                    if (input.Peek() == -1)
                    {
                        break;
                    }
                    else
                    {
                        ch = (char)input.Peek();
                    }
                }
                this.result.Add(int.Parse(accum.ToString()));
            }
            else
            {
                input.Read();

                switch (ch)
                {
                    case '+':
                        this.result.Add(Scanner.Add);
                        break;

                    case '-':
                        this.result.Add(Scanner.Sub);
                        break;

                    case '*':
                        this.result.Add(Scanner.Mul);
                        break;

                    case '/':
                        this.result.Add(Scanner.Div);
                        break;

                    case ';':
                        this.result.Add(Scanner.Semi);
                        break;

                    case '=':
                        if (input.Peek() == '=')
                        {
                            input.Read();
                            this.result.Add(Scanner.Equal);
                        }
                        else
                        {
                            this.result.Add(Scanner.Assign);
                        }
                        break;

                    case '<':
                        if (input.Peek() == '=')
                        {
                            input.Read();
                            this.result.Add(Scanner.LessOrEqual);
                        }
                        else
                        {
                            this.result.Add(Scanner.Less);
                        }
                        break;

                    case '>':
                        if (input.Peek() == '=')
                        {
                            input.Read();
                            this.result.Add(Scanner.MoreOrEqual);
                        }
                        else
                        {
                            this.result.Add(Scanner.More);
                        }
                        break;

                    case '(':
                        this.result.Add(PerenthStart);
                        break;

                    case ')':
                        this.result.Add(PerenthEnd);
                        break;

                    case '{':
                        this.result.Add(BraceEnd);
                        break;

                    case '}':
                        this.result.Add(BraceEnd);
                        break;

                    case '[':
                        this.result.Add(SquareEnd);
                        break;

                    case ']':
                        this.result.Add(SquareEnd);
                        break;

                    default:
                        System.Console.WriteLine("WTF " + ch);
                        break;
                }
            }
        }
    }

}

File: src/Scanner.cpp
#include <sstream>

#include "Scanner.h"
#include "ByteObjects.h"

#define DECLARE_CLASS_ENUM_OBJ(_enum,_value) Object *const Scanner::_value = new _enum(_enum::Enum::_value)

using namespace std;

#pragma region Constants
	DECLARE_CLASS_ENUM_OBJ(BinOp,ADD);
	DECLARE_CLASS_ENUM_OBJ(BinOp,SUB);
	DECLARE_CLASS_ENUM_OBJ(BinOp,MUL);
	DECLARE_CLASS_ENUM_OBJ(BinOp,DIV);
	DECLARE_CLASS_ENUM_OBJ(BinOp,MOD);

	DECLARE_CLASS_ENUM_OBJ(BinOp,SHIFT_LEFT);
	DECLARE_CLASS_ENUM_OBJ(BinOp,SHIFT_RIGHT);

	DECLARE_CLASS_ENUM_OBJ(BinOp,OR);
	DECLARE_CLASS_ENUM_OBJ(BinOp,AND);
	DECLARE_CLASS_ENUM_OBJ(BinOp,XOR);
	DECLARE_CLASS_ENUM_OBJ(BinOp,B_OR);
	DECLARE_CLASS_ENUM_OBJ(BinOp,B_AND);

	DECLARE_CLASS_ENUM_OBJ(LogicOp,NOT);
	DECLARE_CLASS_ENUM_OBJ(LogicOp,B_NOT);

	DECLARE_CLASS_ENUM_OBJ(BinOp,EQUAL);
	DECLARE_CLASS_ENUM_OBJ(BinOp,NOTEQUAL);
	DECLARE_CLASS_ENUM_OBJ(BinOp,MORE);
	DECLARE_CLASS_ENUM_OBJ(BinOp,LESS);
	DECLARE_CLASS_ENUM_OBJ(BinOp,LESS_OR_EQUAL);
	DECLARE_CLASS_ENUM_OBJ(BinOp,MORE_OR_EQUAL);

	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_ADD);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_SUB);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_MUL);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_DIV);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_MOD);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_SHIFT_LEFT);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_SHIFT_RIGHT);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_OR);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_AND);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_XOR);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_B_OR);
	DECLARE_CLASS_ENUM_OBJ(AssignOp,ASSIGN_B_AND);
#pragma endregion

Scanner::Scanner(ifstream &reader)
{
	stringstream ss;
	ss << reader;
	Scan(ss);
}

vector<void *> *Scanner::GetTokens()
{
	return result;
}

void Scanner::Scan(stringstream &reader)
{
	char ch;
		
	while (reader >> ch)
	{
		if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r')
		{
			continue;
		}
		else if (IsLetter(ch))
		{
			stringstream sstream;
			do
			{
				sstream << ch;
			}
			while(reader >> ch && IsLetter(ch));

			string *str;
			sstream >> *str;

			result->push_back(str);
		}
		else if (IsDigit(ch))
		{
			stringstream sstream;
			do
			{
				sstream << ch;
			}
			while(reader >> ch && IsDigit(ch));

			int integer;
			sstream >> integer;

			result->push_back(new Literal<int>(integer));
		}
		else if (ch == '"')
		{
			reader >> ch;
			stringstream sstream;
			sstream << '"';
			do
			{
				sstream << ch;
			}
			while(reader >> ch && ch != '"');

			sstream << '"';
			string str;
			sstream >> str;
			result->push_back(new Literal<string>(str));
			reader >> ch;
		}
		else if (ch == '\'')
		{
			reader >> ch;
			reader >> ch;
			result->push_back(new Literal<char>(ch));
			reader >> ch;
		}
		else switch(ch)
		{
		case '+':	// (ch)= is possible.
		case '-':
		case '*':
		case '/':
		case '%':

		case '|':
		case '&':
		case '^':

		case '!':
		case '~':

		case '=':
		case '>':
		case '<':
			{
				char c;
				reader >> c;
				if (c == '=' ||
					c == ch &&
						(c == '+' || c == '-' || c == '|' || c == '&' ||
						c == '<' || c == '>'))
				{
					char characters[] = {c,ch}; 
					result->push_back(ToArithmethicConstant(string(characters)));
					break;
				}
				else
				{
					reader << c;
					result->push_back(ToArithmethicConstant(ch));
				}
			}
			break;

		case '(':
		case ')':
		case '{':
		case '}':
		case '[':
		case ']':
		case ';':
			result->push_back(ToArithmethicConstant(ch));
			break;
		}
	}
}

bool Scanner::IsLetter(char c)
{
	return c >= 'a' && c <= 'z' ||
		c >= 'A' && c <= 'Z' ||
		c == '_';
}

bool Scanner::IsDigit(char c)
{
	return c >= '0' && c <= '9';
}

void *Scanner::ToArithmethicConstant(char c)
{
	switch(c)
	{
	case '+':
		return Scanner::ADD;
	case '-':
		return Scanner::SUB;
	case '*':
		return Scanner::MUL;
	case '/':
		return Scanner::DIV;
	case '%':
		return Scanner::MOD;

	case '|':
		return Scanner::OR;
	case '&':
		return Scanner::AND;
	case '^':
		return Scanner::XOR;

	case '~':
		return Scanner::NOT;
	case '!':
		return Scanner::B_NOT;

	case '=':
		return Scanner::EQUAL;
	case '>':
		return Scanner::MORE;
	case '<':
		return Scanner::LESS;

	default:
		throw "undefined arithmethic constant '" + string(&c) + "'";
	}
}

void *Scanner::ToArithmethicConstant(string str)
{
	if (str == "||")
	{
		return Scanner::B_OR;
	}
	else if (str == "&&")
	{
		return Scanner::B_AND;
	}
	else if (str == "==")
	{
		return Scanner::EQUAL;
	}
	else if (str == "!=")
	{
		return Scanner::NOTEQUAL;
	}
	else if (str == "<=")
	{
		return Scanner::LESS_OR_EQUAL;
	}
	else if (str == ">=")
	{
		return Scanner::MORE_OR_EQUAL;
	}
	else if (str == "+=")
	{
		return Scanner::ASSIGN_ADD;
	}
	else if (str == "-=")
	{
		return Scanner::ASSIGN_SUB;
	}
	else if (str == "*=")
	{
		return Scanner::ASSIGN_MUL;
	}
	else if (str == "/=")
	{
		return Scanner::ASSIGN_DIV;
	}
	else if (str == "%=")
	{
		return Scanner::ASSIGN_MOD;
	}
	else if (str == "<<=")
	{
		return Scanner::ASSIGN_SHIFT_LEFT;
	}
	else if (str == ">>=")
	{
		return Scanner::ASSIGN_SHIFT_RIGHT;
	}
	else if (str == "|=")
	{
		return Scanner::ASSIGN_OR;
	}
	else if (str == "&=")
	{
		return Scanner::ASSIGN_AND;
	}
	else if (str == "^=")
	{
		return Scanner::ASSIGN_XOR;
	}
	else if (str == "||=")
	{
		return Scanner::ASSIGN_B_OR;
	}
	else if (str == "&&=")
	{
		return Scanner::ASSIGN_B_AND;
	}
	else
	{
		throw "undefined arithmethic constant '" + str + "'";
	}
}